docs/design/2019-11-18-tidb-with-wasm.md

```diff
@@ -0,0 +1,234 @@
+WASI. We know that WebAssembly is an assembly language designed for abstract virtual machines, so it requires 
+a system interface to manipulate the abstracted operating system, which is defined by WASI. With WASI, the 
+application can be compiled once, run anywhere.
+
+Go 1.11 experimentally added the support for WebAssembly and optimized it in Go 1.12. TiDB is written in 
+pure Golang. In theory, it can be compiled directly into WebAssembly according to Golang's official 
+[document](https://github.com/golang/go/wiki/WebAssembly#getting-started), allowing TiDB to run anywhere.
+
+## Implementation
+
+Our implementation will be divided into several stages:
+- Compile TiDB code into Wasm, regardless of whether it can run normally. We should solve the following problem:
+  - TiDB itself or some dependent third-party libraries may not be compatible with Wasm, and the compile will 
+    fail. They need to be modified accordingly.
+- Migrate the result of stage 1 into PC browsers and wrap the page into the TiDB playground and TiDB tour. Possible 
+  problems are:
+  - Golang compatibility with the browser environment.
+  - Due to the browser's security policy, we can't listen to a port using TiDB's original working mode. We need 
+    to integrate a SQL terminal to TiDB instead of connecting via MySQL. At the same time, because the browser 
+    prohibits the application from directly accessing local files, when the user runs  SQL statements related to 
+    file operations, we need to convert the semantics to the upload/download operations in the browser.
+  - TiDB running in a browser cannot execute SQL in batches due to restraints of the browser, but this feature is 
+    useful for users who want to verify TiDB’s compatibility with MySQL or other SQL database. We need to support 
+    it in the browser.
+- Migrate the result of stage 1 into phone browsers. Possible problems are:
+  - The webkit has a [limit](https://trac.webkit.org/browser/webkit/branches/safari-604.1-branch/Source/JavaScriptCore/jit/ExecutableAllocator.cpp#L90) 
+    on executable memory, typically it's 64M. However, the  current Wasm compilation of TiDB  is 76M, which requires 
+    some downsizing.
+  - The browser in phone has different interaction model  than that  in PC, so the SQL terminal must be compatible 
+    with both.
+- Migrate the result of stage 1 into Wasmer, and publish it to [WAPM](https://wapm.io/), so that it can run in any 
+  platform easily with WAPM, even can run on [WebAssembly shell](https://webassembly.sh/), to archive this, we should 
+  solve:
+  - Golang doesn’t support WASI now, so we need to use a [fork branch](https://github.com/neelance/go/tree/wasi) 
+    of Golang and recompile the toolchain.
+  - The result of stage 1 is too large and should be shrinked.
+
+### Solution of compilation problem
+In case of problems of stage 1, incompatibility with WASM can be divided into the following cases:
+
+- TiDB's own code has platform-dependent code
```

Comment:
```suggestion
1. TiDB's own code has platform-dependent code.
```
You use "case 1/2/3/4" to refer to these list items below. In this case, it's recommended to make this list an ordered list.